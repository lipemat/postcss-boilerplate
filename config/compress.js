const config = require( '../helpers/package-config' );

/**
 * Compress files using Brotli.
 *
 * Generates .br versions of the .min files if the package.json
 * has `brotliFiles` set to `true`.
 */
module.exports = {
	brotli: {
		options: {
			mode: 'brotli',
		},
		expand: true,
		cwd: config.css_folder,
		dest: config.css_folder,
		extDot: 'last',
		src: [ '**/*.min.css' ],
		ext: '.css.br'
	}
};
