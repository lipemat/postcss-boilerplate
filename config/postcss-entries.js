const config = require( '../helpers/package-config' );

/**
 * Entry points to be loaded by Grunt.
 *
 * Checks each provided file for exists and includes if does.
 *
 * @see getEntries
 */
module.exports = {
	main: config.file_name, // Default: "front-end".
	admin: 'admin',
	blocks: 'blocks'
};
