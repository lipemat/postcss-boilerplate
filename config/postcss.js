let config = require("../helpers/package-config");
const FileSystemLoader = require( '../lib/FileSystemLoader' );
const modulesFolder = process.env.NODE_ENV === 'production' ? '_css-modules-json/min/' : '_css-modules-json/';

let compileOptions = {
	map: true,
	processors: [
		require('postcss-import')({
			extension: 'pcss',
			plugins : [
				require( 'postcss-modules' )( {
					generateScopedName : process.env.NODE_ENV === 'production' ? '[hash:base64:5]' : '[folder]--[name]__[local]',
					Loader: FileSystemLoader.default,
					globalModulePaths : [
						new RegExp( '.*?' + config.theme_path.replace( /\//g, '\\\\' ) + 'pcss', "i" ),
						new RegExp( '.*?' + config.theme_path + 'pcss', "i" ),
					],
					/**
					 * Custom output of css modules JSON file to specified location
					 * Also excludes json files from the global pcss files
					 */
					getJSON: function(cssFileName, json) {
						const path          = require('path');
						const fse = require('fs-extra');
						let directory = path.relative( config.theme_path, cssFileName ).replace( /\\/g, '/' ) + '/';
						//exclude global pcss
						if ( directory.substr(0,4) === 'pcss' ){
							return;
						}
						let cssName       = path.basename(cssFileName, '.css');
						let jsonFileName  = config.theme_path.replace( /\\/g, '/' ) + modulesFolder + directory.replace( cssName + '/','') + cssName + '.json';

						/**
						 * We use the Sync method here to fix issues where JSON is not
						 * being generated.
						 *
						 * @since 2019-01-22
						 */
						fse.outputJsonSync(jsonFileName, json);
					}
				} ),
			],
		}),
        require('postcss-custom-media'),
        require('postcss-nested'),
        //@link https://preset-env.cssdb.org/features
        //By default all stage 2 features work unless specified otherwise
        require('postcss-preset-env'),
        require('postcss-color-mod-function'),
        require('@lipemat/css-mqpacker'),
	],
	parser: require('postcss-scss'),
};

let minOptions = Object.assign({}, compileOptions);
minOptions.map = false;
minOptions.processors = [ ...compileOptions.processors ];
minOptions.processors.push( require( 'postcss-clean' )( {
		level: 2
	} )
);


module.exports = {
	toCSS : {
		options : compileOptions,
		files : {
			'<%= pkg.theme_path %><%= pkg.css_folder %><%= pkg.file_name %>.css': '<%= pkg.theme_path %>pcss/<%= pkg.file_name %>.pcss'
		},
	},

	min: {
		options: minOptions,
		files: {
			'<%= pkg.theme_path %><%= pkg.css_folder %><%= pkg.file_name %>.min.css': '<%= pkg.theme_path %>pcss/<%= pkg.file_name %>.pcss'
		},
	},
};
