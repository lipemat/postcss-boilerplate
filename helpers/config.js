const packageConfig = require( './package-config' );
const path = require( 'path' );


/**
 * Get a config for our /config directory merged with any
 * matching configuration from the project directory.
 *
 * For instance if we have a file named config/babel.config.js in our project
 * we will merge the contents with our config/babel.config.js in favor of whatever
 * is specified with the project's file.
 *
 * @param {String} fileName
 * @returns {object}
 */
function getConfig( fileName ) {
	let config = require( '../config/' + fileName );
	try {
		let localConfig = require( path.resolve( packageConfig.workingDirectory + '/config', fileName ) );
		config = {...config, ...localConfig};
	} catch ( e ) {
	}
	return config;
}

module.exports = {
	getConfig: getConfig
};
