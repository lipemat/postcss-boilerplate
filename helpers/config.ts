import fs from 'fs';
import path from 'path';
import browserslist from 'browserslist';

import {getPackageConfig} from './package-config';


/**
 * Get a config for our /config directory merged with any
 * matching configuration from the project directory.
 *
 * For instance if we have a file named config/babel.config.js in our project
 * we will merge the contents with our config/babel.config.js in favor of whatever
 * is specified with the project's file.
 */
export function getConfig( fileName: string ) {
	let config = require( '../config/' + fileName );
	const packageConfig = getPackageConfig();
	try {
		const localConfig = require( path.resolve( packageConfig.workingDirectory + '/config', fileName ) );
		if ( 'function' === typeof localConfig ) {
			config = {...config, ...localConfig( config )};
		} else {
			config = {...config, ...localConfig};
		}
	} catch ( e ) {
	}
	return config;
}

/**
 * Get the browserslist from the current project.
 *
 * - If specified using standard browserslist config, we will use that.
 * - Fallback to WordPress defaults except for "> 1%".
 */
export function getBrowsersList() {
	const projectBrowsersList = browserslist();
	if ( browserslist( browserslist.defaults ) === projectBrowsersList ) {
		const wp = [ ...require( '@wordpress/browserslist-config' ) ];
		wp.push( 'not and_uc 15.5' );
		return wp;
	}
	return projectBrowsersList;
}

/**
 * If browserslist is not specified, we fall back to WordPress defaults.
 *
 * Return false if a browserslist is specified in the current project.
 *
 * @deprecated Use getBrowsersList() instead.
 *
 * @link https://github.com/browserslist/browserslist#config-file
 *
 * @return {false | string[]}
 */
export const getDefaultBrowsersList = () => {
	if ( browserslist( browserslist.defaults ) === browserslist() ) {
		const wp = [ ...require( '@wordpress/browserslist-config' ) ];
		wp.push( 'not and_uc 15.5' );
		return wp;
	}
	return false;
};


/**
 * Provide CSS properties and media queries to all postcss plugins.
 *
 * If a media-queries files exist, automatically load them.
 * If CSS variables exist, automatically load them.
 *
 * 1. pcss/globals/variables.pcss
 * 3. pcss/globals/media-queries.pcss
 */
export function getExternalFiles(): string[] {
	const externalFiles: string[] = [];
	const packageConfig = getPackageConfig();

	[
		path.resolve( packageConfig.theme_path, 'pcss/globals/media-queries.pcss' ),
		path.resolve( packageConfig.theme_path, 'pcss/globals/variables.pcss' ),
	].forEach( possibleFile => {
		if ( fs.existsSync( possibleFile ) ) {
			externalFiles.push( possibleFile );
		}
	} );

	return externalFiles;
}
