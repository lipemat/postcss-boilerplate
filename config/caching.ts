import {EnumModules, getEnumFilePath} from '../helpers/enum-modules';
import {JsonModules} from '../helpers/get-json';
import tinylr from 'tiny-lr';
import type {Environment} from '../helpers/config';
import {getPackageConfig} from '../helpers/package-config';


export type CachingGruntTasks = {
	reset: {}
	reload: {}
	writeModules: {}
}

let lastCssClasses: string[] = [];

export const config: CachingGruntTasks = {
	reset: {},
	reload: {},
	writeModules: {},
};

function isEnvironment( env: string | undefined ): env is Environment {
	return 'development' === env || 'production' === env;
}

/**
 * Use the same interface as the grunt watch task
 * to trigger a reload.
 *
 * @link https://github.com/gruntjs/grunt-contrib-watch?tab=readme-ov-file#rolling-your-own-live-reload
 */
function triggerReload() {
	tinylr.changed( getEnumFilePath( 'development' ), () => {
		console.log( 'CSS Modules updated, reloading browser.' );
	} );
}

/**
 * Cache management for Enums and JSON
 *
 * `caching:reset` - Reset the content of the Enums and JSON files.
 * `caching:reload` - Reload the browser if a CSS modules classname changes.
 * `caching:writeModules` - Write contents if JSON or PHP Enums for CSS modules.
 *
 */
export default function cachingTask<K extends keyof CachingGruntTasks>( task: K, env: Environment | undefined ) {
	if ( 'reset' === task ) {
		EnumModules._resetContent();
		JsonModules._resetContent();
	}

	if ( 'reload' === task && getPackageConfig().cssEnums ) {
		if ( lastCssClasses.sort().toString() === EnumModules.getCssClasses().sort().toString() ) {
			return;
		}
		lastCssClasses = EnumModules.getCssClasses();
		triggerReload();
	}

	if ( 'writeModules' === task && isEnvironment( env ) && getPackageConfig().combinedJson ) {
		JsonModules.flushToDisk( env );
		if ( getPackageConfig().cssEnums ) {
			EnumModules.flushToDisk( env );
		}
	}
}
