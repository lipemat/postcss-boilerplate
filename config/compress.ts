import {getPackageConfig} from '../helpers/package-config';

const folder = getPackageConfig().css_folder;

export type CompressGruntTasks = {
	brotli: Config;
}

type Options = {
	pretty?: boolean;
	createEmptyArchive?: boolean;

}

type OptionsZip = {
	level?: number;
	mode: 'zip' | 'gzip';
}

type OptionsBrotli = {
	mode: 'brotli';
	brotli?: { [ constant: number ]: number };

}

type OptionsTar = {
	mode: 'tar' | 'zip' | 'tgz';
	archive?: string | ( () => string );
}


type Config = {
	options: Options | OptionsZip | OptionsBrotli | OptionsTar;
	expand: boolean;
	cwd: string;
	dest: string;
	extDot: 'first' | 'last';
	src: string[];
	ext: string;
};


/**
 * Compress files using Brotli.
 *
 * Generates .br versions of the .min files if the package.json
 * has `brotliFiles` set to `true`.
 */
const config: CompressGruntTasks = {
	brotli: {
		options: {
			mode: 'brotli',
		},
		expand: true,
		cwd: folder,
		dest: folder,
		extDot: 'last',
		src: [ '**/*.min.css' ],
		ext: '.css.br',
	},
};

module.exports = config;
