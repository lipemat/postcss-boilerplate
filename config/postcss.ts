import {getGenerateScopeName} from '../helpers/css-classnames';
import {getPackageConfig} from '../helpers/package-config';
import {getEntries} from '../helpers/entries';
import {type Environment, getBrowsersList, getExternalFiles} from '../helpers/config';
import {getJSON} from '../helpers/get-json';
import type {Plugin, ProcessOptions, Processor} from 'postcss';
import PrettyPlugin from '../lib/postcss-pretty';
import {pluginOptions} from 'postcss-preset-env';
import type {AtImportOptions} from 'postcss-import';

const config = getPackageConfig();
const postcssPresetEnv = require( 'postcss-preset-env' );

export type PostCSSConfig = Pick<ProcessOptions, 'map' | 'parser'> & {
	processors: Plugin[];
	diff?: string| boolean;
	failOnError?: boolean;
	onError?: ( error: Error ) => void;
	sequential?: boolean;
	writeDest?: boolean;
}

export type PostCSSGruntTasks = {
	min: {
		options: PostCSSConfig;
		files: Record<string, string>;
	};
	toCSS: {
		options: PostCSSConfig;
		files: Record<string, string>;
	};
}

/**
 * Base postcss-presets-env config.
 *
 */
const presetEnv: pluginOptions = {
	browsers: [ ...getBrowsersList() ],
	features: {},
};

// Get a list of included postcss plugins based no the browsers list.
const includedPlugins: string[] = postcssPresetEnv( presetEnv ).plugins.map( ( plugin: Processor['plugins'][number] ) => {
	return 'postcssPlugin' in plugin ? plugin.postcssPlugin : '';
} );


if ( 'object' === typeof presetEnv.features && includedPlugins.includes( 'postcss-focus-visible' ) ) {
	presetEnv.features[ 'focus-visible-pseudo-class' ] = {
		/**
		 * Fixes `focus-visible` feature for CSS modules.
		 *
		 * Only needed if our browsers list includes non-supported browsers
		 * such as Safari 15.3 and below.
		 *
		 * Requires `focus-visible` polyfill to be loaded externally.
		 * Most will often need it site wide on pages, which do and don't use the JS app.
		 *
		 * @link https://unpkg.com/focus-visible@5.2.0/dist/focus-visible.min.js
		 */
		replaceWith: ':global(.focus-visible)',
	};
}

/**
 * A reusable config for postcss-import based on the environment.
 *
 * @param {'production'|'development'} env The environment to use.
 */
function getImportConfig( env: Environment ): AtImportOptions {
	return {
		plugins: [
			require( 'postcss-modules' )( {
				generateScopedName: getGenerateScopeName( env ),
				globalModulePaths: [
					new RegExp( '.*?' + config.theme_path.replace( /\//g, '\\\\' ) + 'pcss', 'i' ),
					new RegExp( '.*?' + config.theme_path + 'pcss', 'i' ),
				],
				getJSON: getJSON( env ),
			} ),
		],
		skipDuplicates: false,
	};
}


const compileOptions: PostCSSConfig = {
	map: true,
	processors: [
		require( '@csstools/postcss-global-data' )( {
			files: getExternalFiles(),
		} ),
		require( 'postcss-import' )( getImportConfig( 'development' ) ),
		require( 'postcss-custom-media' ),
		require( 'postcss-nested' ),
		postcssPresetEnv( presetEnv ),
		require( 'postcss-color-mod-function' ),
		require( 'postcss-sort-media-queries' )( {
			onlyTopLevel: true,
			sort: 'mobile-first',
			configuration: {
				unitlessMqAlwaysFirst: true,
			},
		} ),
		// Create a manifest for browser cache flushing.
		require( 'postcss-hash' )( {
			algorithm: 'md5',
			trim: 20,
			manifest: config.css_folder + '/manifest.json',
			name: ( {hash} ) => hash,
		} ),
	],
	parser: require( 'postcss-scss' ),
};

const minOptions = Object.assign( {}, compileOptions );
minOptions.map = false;
minOptions.processors = [ ...compileOptions.processors ].map( processor => {
	if ( 'postcss-import' === processor.postcssPlugin ) {
		return require( 'postcss-import' )( getImportConfig( 'production' ) );
	}
	return processor;
} );
minOptions.processors.push( require( '../lib/postcss-clean' )( {
	level: 2,
} ) );

// Add pretty output for development.
compileOptions.processors.push( PrettyPlugin );

const gruntTasks: PostCSSGruntTasks = {
	toCSS: {
		options: compileOptions,
		files: getEntries().toCSS,
	},
	min: {
		options: minOptions,
		files: getEntries().min,
	},
};

module.exports = gruntTasks;
