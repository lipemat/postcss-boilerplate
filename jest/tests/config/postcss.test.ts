import {readFileSync} from 'fs';
import postcss, {Plugin} from 'postcss';
import {basename} from 'path';
import {adjustBrowserslist} from '../../../helpers/config';
import type {PostCSSGruntTasks} from '../../../config/postcss';

const browserslist = require( 'browserslist' );
const postcssPresetEnv = require( 'postcss-preset-env' );

export type Fixture = {
	input: string;
	output: string;
	basename: string;
	description: string;
}

const creator = ( browsers: string[], features = {} ) => {
	return postcssPresetEnv( {
		browsers,
		features: {...features},
	} );
};


function getPostCSSConfig(): PostCSSGruntTasks {
	// @ts-ignore
	let config: PostCSSGruntTasks = {};
	jest.isolateModules( () => {
		config = require( '../../../config/postcss' );
	} );
	return config;
}


function processPostCSS( input: string, min: boolean = false, file: string ): Promise<postcss.Result> {
	const config = getPostCSSConfig();
	const task = min ? 'min' : 'toCSS';
	const plugins = config[ task ].options.processors;

	// Prevent the manifest.json from being emitted.
	const hashIndex = plugins.findIndex( ( plugin: Plugin ) => {
		return plugin.postcssPlugin !== undefined && 'postcss-hash' === plugin.postcssPlugin;
	} );
	if ( hashIndex > -1 ) {
		plugins.splice( hashIndex, 1 );
	}

	return postcss( plugins ).process( input, {
		from: file,
		to: 'test.css',
		parser: config[ task ].options.parser,
	} );
}

function cleanCSS( css: string ): string {
	return css.replace( /undefined/g, '' ).trim();
}

function getBrowsersPlugin( plugins: Plugin[] ): { plugins: Plugin[] } {
	return plugins.find( plugin => 'postcss-preset-env' === plugin.postcssPlugin ) as unknown as {
		plugins: Plugin[]
	};
}

// Create a data provider for fixtures.
const fixtures: Fixture[] = require( 'glob' )
	.sync( 'jest/fixtures/{postcss,safari-15}/*.pcss' )
	.map( file => {
		return {
			basename: basename( file ),
			input: file,
			output: file.replace( '.pcss', '.css' ),
			description: file.replace( /\\/g, '/' ).replace( 'jest/fixtures/', '' ),
		};
	} );


afterEach( () => {
	delete process.env.BROWSERSLIST;
} );


describe( 'postcss.js', () => {
	test( 'PostCSS config', () => {
		/**
		 * @notice If the browserslist results change, the snapshot will need to be updated.
		 */
		const config = require( '../../../config/postcss' );
		expect( config ).toMatchSnapshot( 'develop' );
	} );


	test( 'sourceMap', () => {
		const config = require( '../../../config/postcss' );
		expect( config.toCSS.options.map ).toEqual( true );
		expect( config.min.options.map ).toEqual( false );
	} );


	test( 'Browserslist config', () => {
		const expectedBrowsers = adjustBrowserslist( [ ...require( '@wordpress/browserslist-config' ) ] );

		const config = getPostCSSConfig();
		// We want to make sure no matter what postcss-custom-properties is not included
		// if a user did not provided a custom browserslist to override.
		expect( getBrowsersPlugin( config.toCSS.options.processors ).plugins.filter( plugin => {
			return 'postcss-custom-properties' === plugin.postcssPlugin;
		} ).length ).toEqual( 0 );
		expect( getBrowsersPlugin( config.toCSS.options.processors ).plugins.filter( plugin => {
			return 'postcss-focus-visible' === plugin.postcssPlugin;
		} ).length ).toEqual( 0 );

		expect( JSON.stringify( getBrowsersPlugin( config.toCSS.options.processors ) ) )
			.toEqual( JSON.stringify( creator( expectedBrowsers ) ) );
		expect( JSON.stringify( getBrowsersPlugin( config.min.options.processors ) ) )
			.toEqual( JSON.stringify( creator( expectedBrowsers ) ) );

		// op_mini all requires postcss-custom-properties.
		process.env.BROWSERSLIST = 'op_mini all';
		const config2 = getPostCSSConfig();
		expect( getBrowsersPlugin( config2.toCSS.options.processors ).plugins.filter( plugin => {
			return 'postcss-custom-properties' === plugin.postcssPlugin;
		} ).length ).toEqual( 1 );
		expect( JSON.stringify( getBrowsersPlugin( config2.toCSS.options.processors ) ) )
			.toEqual( JSON.stringify( creator( [ 'op_mini all' ] ) ) );
		expect( JSON.stringify( getBrowsersPlugin( config2.min.options.processors ) ) )
			.toEqual( JSON.stringify( creator( [ 'op_mini all' ] ) ) );

		// Safari 15 requires postcss-focus-visible.
		process.env.BROWSERSLIST = 'safari 15';
		const config4 = getPostCSSConfig();
		expect( getBrowsersPlugin( config4.toCSS.options.processors ).plugins.filter( plugin => {
			return 'postcss-focus-visible' === plugin.postcssPlugin;
		} ).length ).toEqual( 1 );

		expect( JSON.stringify( getBrowsersPlugin( config4.toCSS.options.processors ) ) )
			.toEqual( JSON.stringify( creator( [ 'safari 15' ], {
				'focus-visible-pseudo-class': {
					replaceWith: ':global(.focus-visible)',
				},
			} ) ) );

		// @notice If this fails, we can probably remove the override from adjustBrowserslist and the toEqual to `0`.
		const wpDefaultBrowsers = [ ...require( '@wordpress/browserslist-config' ) ];
		process.env.BROWSERSLIST = browserslist( wpDefaultBrowsers );
		expect( getBrowsersPlugin( getPostCSSConfig().toCSS.options.processors ).plugins.filter( plugin => {
			return 'postcss-custom-properties' === plugin.postcssPlugin;
		} ).length ).toEqual( 1 );
	} );

	test.each( fixtures )( 'PostCSS fixtures ( $description )', async fixture => {
		if ( fixture.input.includes( 'safari-15' ) ) {
			process.env.BROWSERSLIST = 'safari 15';
		}

		const input = readFileSync( fixture.input, 'utf8' );
		let output = readFileSync( fixture.output.replace( '.css', '.raw.css' ), 'utf8' );
		let result = await processPostCSS( input, false, fixture.input );
		expect( cleanCSS( result.css ) ).toEqual( output.trim() );

		result = await processPostCSS( input, true, fixture.input );
		output = readFileSync( fixture.output.replace( '.css', '.raw.min.css' ), 'utf8' );
		expect( cleanCSS( result.css ) ).toEqual( output.trim() );
	} );
} );
