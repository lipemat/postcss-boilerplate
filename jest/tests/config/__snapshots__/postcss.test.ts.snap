// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`postcss.js PostCSS config: develop 1`] = `
{
  "min": {
    "files": {},
    "options": {
      "map": false,
      "parser": {
        "parse": [Function],
        "stringify": [Function],
      },
      "processors": [
        {
          "postcssPlugin": "postcss-global-data",
          "prepare": [Function],
        },
        {
          "Once": [Function],
          "postcssPlugin": "postcss-import",
        },
        [Function],
        [Function],
        {
          "plugins": [
            {
              "Declaration": [Function],
              "postcssPlugin": "postcss-color-function",
            },
            {
              "postcssPlugin": "postcss-normalize-display-values",
              "prepare": [Function],
            },
            {
              "AtRule": [Function],
              "postcssPlugin": "postcss-font-format-keywords",
            },
            {
              "Declaration": [Function],
              "postcssPlugin": "postcss-oklab-function",
            },
            {
              "postcssPlugin": "postcss-text-decoration-shorthand",
              "prepare": [Function],
            },
            {
              "AtRule": [Function],
              "Once": [Function],
              "postcssPlugin": "postcss-custom-media",
            },
            {
              "Rule": [Function],
              "postcssPlugin": "postcss-nesting",
            },
            {
              "Rule": [Function],
              "postcssPlugin": "postcss-selector-not",
            },
            {
              "Rule": [Function],
              "postcssPlugin": "postcss-dir-pseudo-class",
            },
            {
              "Declaration": [Function],
              "postcssPlugin": "postcss-lab-function",
            },
            {
              "Rule": [Function],
              "postcssPlugin": "css-blank-pseudo",
            },
            {
              "Declaration": [Function],
              "postcssPlugin": "postcss-page-break",
            },
            {
              "Once": [Function],
              "postcssPlugin": "postcss-font-variant",
            },
            {
              "Rule": [Function],
              "postcssPlugin": "css-has-pseudo",
            },
            {
              "Declaration": [Function],
              "postcssPlugin": "postcss-stepped-value-functions",
            },
            {
              "Declaration": [Function],
              "postcssPlugin": "postcss-trigonometric-functions",
            },
            {
              "browsers": [
                "> 1%",
                "last 1 Android versions",
                "last 1 ChromeAndroid versions",
                "last 2 Chrome versions",
                "last 2 Firefox versions",
                "last 2 Safari versions",
                "last 2 iOS versions",
                "last 2 Edge versions",
                "last 2 Opera versions",
                "not op_mini all",
              ],
              "info": [Function],
              "options": {
                "overrideBrowserslist": [
                  "> 1%",
                  "last 1 Android versions",
                  "last 1 ChromeAndroid versions",
                  "last 2 Chrome versions",
                  "last 2 Firefox versions",
                  "last 2 Safari versions",
                  "last 2 iOS versions",
                  "last 2 Edge versions",
                  "last 2 Opera versions",
                  "not op_mini all",
                ],
              },
              "postcssPlugin": "autoprefixer",
              "prepare": [Function],
            },
            {
              "RuleExit": [Function],
              "postcssPlugin": "postcss-progressive-custom-properties",
            },
            {
              "OnceExit": [Function],
              "postcssPlugin": "postcss-preset-env",
            },
          ],
          "postcssPlugin": "postcss-preset-env",
        },
        [Function],
        [Function],
        {
          "OnceExit": [Function],
          "postcssPlugin": "postcss-hash",
        },
        {
          "OnceExit": [Function],
          "postcssPlugin": "clean",
        },
      ],
    },
  },
  "toCSS": {
    "files": {},
    "options": {
      "map": true,
      "parser": {
        "parse": [Function],
        "stringify": [Function],
      },
      "processors": [
        {
          "postcssPlugin": "postcss-global-data",
          "prepare": [Function],
        },
        {
          "Once": [Function],
          "postcssPlugin": "postcss-import",
        },
        [Function],
        [Function],
        {
          "plugins": [
            {
              "Declaration": [Function],
              "postcssPlugin": "postcss-color-function",
            },
            {
              "postcssPlugin": "postcss-normalize-display-values",
              "prepare": [Function],
            },
            {
              "AtRule": [Function],
              "postcssPlugin": "postcss-font-format-keywords",
            },
            {
              "Declaration": [Function],
              "postcssPlugin": "postcss-oklab-function",
            },
            {
              "postcssPlugin": "postcss-text-decoration-shorthand",
              "prepare": [Function],
            },
            {
              "AtRule": [Function],
              "Once": [Function],
              "postcssPlugin": "postcss-custom-media",
            },
            {
              "Rule": [Function],
              "postcssPlugin": "postcss-nesting",
            },
            {
              "Rule": [Function],
              "postcssPlugin": "postcss-selector-not",
            },
            {
              "Rule": [Function],
              "postcssPlugin": "postcss-dir-pseudo-class",
            },
            {
              "Declaration": [Function],
              "postcssPlugin": "postcss-lab-function",
            },
            {
              "Rule": [Function],
              "postcssPlugin": "css-blank-pseudo",
            },
            {
              "Declaration": [Function],
              "postcssPlugin": "postcss-page-break",
            },
            {
              "Once": [Function],
              "postcssPlugin": "postcss-font-variant",
            },
            {
              "Rule": [Function],
              "postcssPlugin": "css-has-pseudo",
            },
            {
              "Declaration": [Function],
              "postcssPlugin": "postcss-stepped-value-functions",
            },
            {
              "Declaration": [Function],
              "postcssPlugin": "postcss-trigonometric-functions",
            },
            {
              "browsers": [
                "> 1%",
                "last 1 Android versions",
                "last 1 ChromeAndroid versions",
                "last 2 Chrome versions",
                "last 2 Firefox versions",
                "last 2 Safari versions",
                "last 2 iOS versions",
                "last 2 Edge versions",
                "last 2 Opera versions",
                "not op_mini all",
              ],
              "info": [Function],
              "options": {
                "overrideBrowserslist": [
                  "> 1%",
                  "last 1 Android versions",
                  "last 1 ChromeAndroid versions",
                  "last 2 Chrome versions",
                  "last 2 Firefox versions",
                  "last 2 Safari versions",
                  "last 2 iOS versions",
                  "last 2 Edge versions",
                  "last 2 Opera versions",
                  "not op_mini all",
                ],
              },
              "postcssPlugin": "autoprefixer",
              "prepare": [Function],
            },
            {
              "RuleExit": [Function],
              "postcssPlugin": "postcss-progressive-custom-properties",
            },
            {
              "OnceExit": [Function],
              "postcssPlugin": "postcss-preset-env",
            },
          ],
          "postcssPlugin": "postcss-preset-env",
        },
        [Function],
        [Function],
        {
          "OnceExit": [Function],
          "postcssPlugin": "postcss-hash",
        },
        [Function],
      ],
    },
  },
}
`;
